### ISSUE TITLE
Auto-configuration for spring-security-oauth2-resource-server by oidc issuer location

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-boot#14150]
Spring Boot should add auto configuration support by the oidc issuer location for spring-security-oauth2-resource-server.

For example, a user could configure a resource server in the following way:

```yaml
spring:
  security:
    oauth2:
      resource-server:
        jwt:
          oidc-issuer-location: https://example.org
```

Then, if spring-security-oauth2-resource-server is present on the classpath, as attested by the presence of `org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken`, Spring Boot's auto configuration support could configure a resource server in the following way:

```java
http
    .authorizeRequests()
        .anyRequest().authenticated()
        .and()
    .oauth2()
        .resourceServer()
            .jwt();
```

And then could expose an instance of `JwtDecoder` by invoking:

```java
JwtDecoders.fromOidcIssuerLocation(oidcIssuerLocation);
```

The main value of this specific name is that is comes directly from [the spec language](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig):

> Using the Issuer location discovered as described in Section 2 or by other means, the OpenID Provider's configuration information can be retrieved.

If the user is curious, she can google "oidc issuer location", and the above spec is the first  link, which is also helpful.

It also clarifies the semantics of the value since there is more than one OAuth spec for discovering configuration, e.g. https://tools.ietf.org/html/rfc8414

Also, while of marginal benefit, it is nice that it matches the method name.

**Comment by jgrandja** [comment]
The updated DSL should be:

```java
http.oauth2ResourceServer().jwt();
```

See related issue [#5715](https://github.com/spring-projects/spring-security/issues/5715)

**Comment by ayudovin** [comment]
I begin working on this enhancement and now I'm blocked by adding unit test because this method `JwtDecoders.fromOidcIssuerLocation(oidcIssuerLocation)` sends request via `restTemplate` :
`return rest.getForObject(oidcIssuerLocation + "/.well-known/openid-configuration", String.class);`.
Could you give some advice how to mock it ?

**Comment by mbhave** [comment]
@ayudovin you can look into `MockWebServer`. There is an example [here](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/security/oauth2/client/OAuth2ClientPropertiesRegistrationAdapterTests.java#L308).

**Comment by rwinch** [comment]
I agree with @mbhave `MockWebServer` is a great approach if you need to use the actual `JwtDecoder` that is created from the OIDC Issuer Location. 

To add to that...an alternative is to use `@MockBean JwtDecoder` and you can then mock the Jwt that is returned. Yet another option is to use Spring Security [test support](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#test) to [run as an authenticated user](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#test-mockmvc-securitycontextholder).

**Comment by mbhave** [comment]
FWIW, I don't think we'd want to mock the `JwtDecoder` because we need to test that the auto-configuration creates the `JwtDecoder` bean using `JwtDecoders.fromOidcIssuerLocation(oidcIssuerLocation);`.

**Comment by rwinch** [comment]
Sorry I wasn't clear. Agreed. If you are wanting to test `JwtDecoders.fromOidcIssuerLocation`, then `MockWebServer` is the right approach. The alternatives are for scenarios where you are trying to test other aspects of an application.

**Comment by mbhave** [comment]
Closing in favor of PR #14190.
